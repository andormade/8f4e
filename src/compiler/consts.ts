export const I32_SIGNED_SMALLEST_NUMBER = -2147483648;
