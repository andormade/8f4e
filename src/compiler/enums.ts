export const enum NameSection {
	FUNCTION_NAME = 0x01,
	LOCAL_NAME = 0x02,
}

export const enum ImportDesc {
	MEMORY = 0x02,
}

export const enum ExportDesc {
	FUNC = 0x00,
}

export const enum Type {
	I32 = 0x7f,
	F32 = 0x7d,
}

export const enum Instruction {
	IF = 0x04,
	ELSE = 0x05,
	END = 0x0b,
	CALL = 0x10,
	LOCAL_GET = 0x20,
	I32_LOAD = 0x28,
	F32_LOAD = 0x2a,
	I32_STORE = 0x36,
	I32_CONST = 0x41,
	F32_CONST = 0x43,
	I32_EQ = 0x46,
	I32_GE_S = 0x4e,
	I32_ADD = 0x6a,
	I32_SUB = 0x6b,
	I32_MUL = 0x6c,
	I32_DIV_U = 0x6e,
	F32_STORE = 0x83,
	F32_ADD = 0x92,
	F32_SUB = 0x93,
	F32_MUL = 0x94,
	F32_DIV = 0x95,
}

export const enum Section {
	CUSTOM = 0x00,
	TYPE = 0x01,
	IMPORT = 0x02,
	FUNCTION = 0x03,
	MEMORY = 0x05,
	EXPORT = 0x07,
	CODE = 0x0a,
}
