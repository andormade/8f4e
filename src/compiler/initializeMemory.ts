const generateOutputAddressLookup = function (compiledModules) {
	const lookup = {};
	compiledModules.forEach(({ outputs, moduleId }) => {
		lookup[moduleId] = outputs.map(output => ({ outputId: output.id, memoryAddress: output.address }));
	});
	return lookup;
};

export const setInitialMemory = function (memory: any, initialMemory: any) {
	for (let i = 0; i < initialMemory.length; i++) {
		memory[i] = initialMemory[i];
	}
};

export const initializeMemory = function (compiledModules, connections) {
	const memoryRef = new WebAssembly.Memory({ initial: 1 });
	const memoryBuffer = new Int32Array(memoryRef.buffer);

	const initialMemory = compiledModules.map(({ initialMemory }) => initialMemory).flat();

	setInitialMemory(memoryBuffer, initialMemory);

	const outputAddressLookup = generateOutputAddressLookup(compiledModules);

	return { memoryRef, memoryBuffer, outputAddressLookup };
};
