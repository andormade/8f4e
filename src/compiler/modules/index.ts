export { default as attenuator } from './attenuator';
export { default as randomGenerator } from './randomGenerator';
export { default as saw } from './saw';
export { default as splitter } from './splitter';
export { default as clockGenerator } from './clockGenerator';
export { default as quantizer } from './quantizer';
