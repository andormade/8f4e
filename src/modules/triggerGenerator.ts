import singleSliderModule from './templates/singleSliderModule';
import source, { extractState, insertState } from './engines/triggerGenerator.asm';
import addDefaultInputPositions from './helpers/addDefaultInputPositions';

import { ModuleType, SliderChangeHandler } from '../state/types';

const onChange: SliderChangeHandler = function (module, memoryBuffer, memoryAddressLookup, movement, slider) {
	let { rate } = extractState(memoryBuffer, memoryAddressLookup[module.id].__startAddress);
	rate = Math.min(Math.max(slider.minValue, rate + movement * -1 * slider.resolution), slider.maxValue);
	insertState({ rate }, memoryBuffer, memoryAddressLookup[module.id].__startAddress);
};

export default function triggerGenerator(): ModuleType {
	return {
		...singleSliderModule({
			id: 'rate',
			maxValue: 100,
			minValue: 1,
			resolution: 1,
			onChange,
		}),
		buttons: [],
		category: 'Oscillator',
		engine: { source },
		initialState: { rate: 10 },
		inputs: addDefaultInputPositions([{ id: 'reset' }]),
		name: 'Trigger',
		extractState,
		insertState,
	};
}
