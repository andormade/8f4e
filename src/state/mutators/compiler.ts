import { State } from '../types';

export default function compiler(state: State, events) {
	const worker = new Worker(new URL('../../../packages/worker/src/index.ts', import.meta.url));
	// @ts-ignore shared: true
	const memoryRef = new WebAssembly.Memory({ initial: 1, maximum: 1, shared: true });

	async function recompile() {
		worker.postMessage({ memoryRef, modules: state.modules, connections: state.connections });
		state.compiler.isCompiling = true;
		state.compiler.lastCompilationStart = performance.now();
	}

	async function onWorkerMessage({ data }) {
		switch (data.type) {
			case 'midiMessage':
				events.dispatch('sendMidiMessage', data.payload);
				break;
			case 'compilationDone':
				state.compiler.memoryBuffer = new Int32Array(memoryRef.buffer);
				state.compiler.outputAddressLookup = data.payload.outputAddressLookup;
				state.compiler.isCompiling = false;
				const end = performance.now() - state.compiler.lastCompilationStart;
				state.compiler.compilationTime = end.toFixed(2);

				state.modules.forEach(module => {
					if (state.moduleTypes[module.type].transformer) {
						state.moduleTypes[module.type].transformer(
							module,
							state.compiler.memoryBuffer,
							state.compiler.outputAddressLookup
						);
					}
				});
				break;
		}
	}

	worker.addEventListener('message', onWorkerMessage);
	events.on('createConnection', recompile);
	events.on('deleteConnection', recompile);
	events.on('addModule', recompile);
	events.on('deleteModule', recompile);
	events.on('init', recompile);

	return () => {
		events.off('createConnection', recompile);
		events.off('deleteConnection', recompile);
		events.off('addModule', recompile);
		events.off('deleteModule', recompile);
		events.off('init', recompile);
	};
}
