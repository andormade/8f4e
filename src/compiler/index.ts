import {
	createFunctionSection,
	createTypeSection,
	createFunctionType,
	createExportSection,
	createFunctionExport,
	createCodeSection,
	createFunctionBody,
	createImportSection,
	createMemoryImport,
} from './wasm/sections';

import { call, i32store } from './wasm/instructions';
import { generateOutputAddressLookup } from './initializeMemory';
import * as moduleCompilers from './modules';
import { Module } from './modules/types';
import { i32abs, i32modulo } from './wasm/helpers';
import { Type } from 'wasm-bytecode-utils';

const HEADER = [0x00, 0x61, 0x73, 0x6d];
const VERSION = [0x01, 0x00, 0x00, 0x00];

const compileModules = function (modules): Module[] {
	let memoryAddress = 4;
	return modules
		.filter(({ type }) => moduleCompilers[type])
		.map(({ id, type }) => {
			const module = moduleCompilers[type](id, memoryAddress);
			memoryAddress += module.initialMemory.length * Int32Array.BYTES_PER_ELEMENT;
			return module;
		});
};

const generateMemoryInitiatorFunction = function (compiledModules) {
	return compiledModules
		.map(module => {
			let pointer = module.offset;
			return module.initialMemory
				.map(value => {
					const instuction = i32store(pointer, value);
					pointer += Int32Array.BYTES_PER_ELEMENT;
					return instuction;
				})
				.flat();
		})
		.flat();
};

const compile = function (modules: object[]) {
	const compiledModules = compileModules(modules);
	const functionBodies = compiledModules.map(({ functionBody }) => functionBody);
	const functionSignatures = compiledModules.map(() => 0x00);
	const cycleFunction = compiledModules.map((module, index) => call(index + 4)).flat();
	const memoryInitiatorFunction = generateMemoryInitiatorFunction(compiledModules);
	const outputAddressLookup = generateOutputAddressLookup(compiledModules);
	const helperFunctions = [i32abs(), i32modulo()];
	const helperFunctionSignatures = [0x01, 0x02];

	return {
		codeBuffer: Uint8Array.from([
			...HEADER,
			...VERSION,
			...createTypeSection([
				createFunctionType([], []),
				createFunctionType([Type.I32], [Type.I32]),
				createFunctionType([Type.I32, Type.I32], [Type.I32]),
			]),
			...createImportSection([createMemoryImport('js', 'memory', 1, 1, true)]),
			...createFunctionSection([...helperFunctionSignatures, 0x00, 0x00, ...functionSignatures]),
			...createExportSection([createFunctionExport('init', 0x02), createFunctionExport('cycle', 0x03)]),
			...createCodeSection([
				...helperFunctions,
				createFunctionBody([], memoryInitiatorFunction),
				createFunctionBody([], cycleFunction),
				...functionBodies,
			]),
		]),
		outputAddressLookup,
		compiledModules,
	};
};

export default compile;
