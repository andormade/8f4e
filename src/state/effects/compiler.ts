import { State } from '../types';
import { compilationDone, recompile } from '../mutators/compiler';

export default function compiler(state: State, events): void {
	const worker = new Worker(new URL('../../../packages/worker/src/index.ts', import.meta.url));
	// @ts-ignore shared: true
	const memoryRef = new WebAssembly.Memory({ initial: 1, maximum: 1, shared: true });

	async function onRecompile() {
		worker.postMessage({ memoryRef, modules: state.modules, connections: state.connections });
		recompile(state);
	}

	async function onWorkerMessage({ data }) {
		switch (data.type) {
			case 'midiMessage':
				events.dispatch('sendMidiMessage', data.payload);
				break;
			case 'compilationDone':
				compilationDone(state, data, memoryRef);
				state.modules.forEach(({ type }, index) => {
					if (state.moduleTypes[type].insertState) {
						state.moduleTypes[type].insertState(
							state.modules[index].state,
							state.compiler.memoryBuffer,
							state.compiler.memoryAddressLookup[state.modules[index].id]
						);
					}
				});
				break;
		}
	}

	worker.addEventListener('message', onWorkerMessage);
	events.on('createConnection', onRecompile);
	events.on('deleteConnection', onRecompile);
	events.on('addModule', onRecompile);
	events.on('deleteModule', onRecompile);
	events.on('init', onRecompile);
}
