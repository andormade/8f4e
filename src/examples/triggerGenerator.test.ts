import { createTestModule, I16_SIGNED_LARGEST_NUMBER, TestModule } from '@8f4e/compiler';

import triggerGenerator from './triggerGenerator.asm';

let testModule: TestModule;

describe('triggerGenerator module', () => {
	beforeAll(async () => {
		testModule = await createTestModule(triggerGenerator);
	});

	beforeEach(() => {
		testModule.reset();
	});

	test('if the generated ast matches with the snapshot', () => {
		expect(testModule.ast).toMatchSnapshot();
	});

	test('if the wat code matches with the snapshot', () => {
		expect(testModule.wat).toMatchSnapshot();
	});

	test('if the generated memory map matches with the snapshot', () => {
		expect(testModule.memoryMap).toMatchSnapshot();
	});

	test('if the counter is counting', () => {
		const { memory, test } = testModule;

		memory.set('rate', 10);

		for (let i = 0; i < 30; i++) {
			expect(memory.get('counter')).toBe(i % 11);
			test();
		}
	});

	test('if the pulse is coming', () => {
		const { memory, test } = testModule;

		memory.set('rate', 1);

		for (let i = 0; i < 10; i++) {
			test();
			expect(memory.get('out')).toBe(0);
			test();
			expect(memory.get('out')).toBe(I16_SIGNED_LARGEST_NUMBER);
		}
	});

	test('if the pulse is coming', () => {
		const { memory, test } = testModule;

		memory.set('rate', 10);

		for (let i = 0; i < 10; i++) {
			for (let j = 0; j < 10; j++) {
				test();
				expect(memory.get('out')).toBe(0);
			}
			test();
			expect(memory.get('out')).toBe(I16_SIGNED_LARGEST_NUMBER);
		}
	});

	test('reset input', () => {
		const { memory, test } = testModule;
		memory.set('rate', 10);

		for (let j = 0; j < 10; j++) {
			test();
			expect(memory.get('out')).toBe(0);
		}

		memory.set('reset', 1); // Send reset signal
		test();

		for (let j = 0; j < 10; j++) {
			test();
			expect(memory.get('out')).toBe(0);
		}
	});
});
