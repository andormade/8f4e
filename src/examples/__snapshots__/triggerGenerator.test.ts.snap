// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`triggerGenerator module if the generated ast matches with the snapshot 1`] = `
Array [
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "triggerGenerator",
      },
    ],
    "instruction": "module",
    "lineNumber": 2,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "counter",
      },
      Object {
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "private",
    "lineNumber": 4,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "rate",
      },
      Object {
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "public",
    "lineNumber": 5,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "out",
      },
      Object {
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "output",
    "lineNumber": 6,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "defaultValue",
      },
      Object {
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "private",
    "lineNumber": 7,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "reset",
      },
      Object {
        "type": "identifier",
        "value": "defaultValue",
      },
    ],
    "instruction": "inputPointer",
    "lineNumber": 8,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_counter",
      },
    ],
    "instruction": "local",
    "lineNumber": 10,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_output",
      },
    ],
    "instruction": "local",
    "lineNumber": 11,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_rate",
      },
    ],
    "instruction": "local",
    "lineNumber": 12,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "out",
      },
    ],
    "instruction": "push",
    "lineNumber": 14,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_output",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 15,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "counter",
      },
    ],
    "instruction": "push",
    "lineNumber": 17,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_counter",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 18,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "rate",
      },
    ],
    "instruction": "push",
    "lineNumber": 20,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_rate",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 21,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "reset",
      },
    ],
    "instruction": "push",
    "lineNumber": 23,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "push",
    "lineNumber": 24,
  },
  Object {
    "arguments": Array [],
    "instruction": "greaterThan",
    "lineNumber": 25,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "void",
      },
    ],
    "instruction": "if",
    "lineNumber": 26,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "push",
    "lineNumber": 27,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_counter",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 28,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "push",
    "lineNumber": 29,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_output",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 30,
  },
  Object {
    "arguments": Array [],
    "instruction": "end",
    "lineNumber": 31,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_output",
      },
    ],
    "instruction": "push",
    "lineNumber": 34,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 32767,
      },
    ],
    "instruction": "push",
    "lineNumber": 35,
  },
  Object {
    "arguments": Array [],
    "instruction": "greaterOrEqual",
    "lineNumber": 36,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "void",
      },
    ],
    "instruction": "if",
    "lineNumber": 37,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "push",
    "lineNumber": 38,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_output",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 39,
  },
  Object {
    "arguments": Array [],
    "instruction": "end",
    "lineNumber": 40,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_counter",
      },
    ],
    "instruction": "push",
    "lineNumber": 42,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_rate",
      },
    ],
    "instruction": "push",
    "lineNumber": 43,
  },
  Object {
    "arguments": Array [],
    "instruction": "greaterOrEqual",
    "lineNumber": 44,
  },
  Object {
    "arguments": Array [],
    "instruction": "if",
    "lineNumber": 45,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 32767,
      },
    ],
    "instruction": "push",
    "lineNumber": 46,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_output",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 47,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "push",
    "lineNumber": 49,
  },
  Object {
    "arguments": Array [],
    "instruction": "else",
    "lineNumber": 50,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_counter",
      },
    ],
    "instruction": "push",
    "lineNumber": 51,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "literal",
        "value": 1,
      },
    ],
    "instruction": "push",
    "lineNumber": 52,
  },
  Object {
    "arguments": Array [],
    "instruction": "add",
    "lineNumber": 53,
  },
  Object {
    "arguments": Array [],
    "instruction": "end",
    "lineNumber": 54,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_counter",
      },
    ],
    "instruction": "localSet",
    "lineNumber": 56,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "&out",
      },
    ],
    "instruction": "push",
    "lineNumber": 58,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_output",
      },
    ],
    "instruction": "push",
    "lineNumber": 59,
  },
  Object {
    "arguments": Array [],
    "instruction": "store",
    "lineNumber": 60,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "&counter",
      },
    ],
    "instruction": "push",
    "lineNumber": 62,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "_counter",
      },
    ],
    "instruction": "push",
    "lineNumber": 63,
  },
  Object {
    "arguments": Array [],
    "instruction": "store",
    "lineNumber": 64,
  },
]
`;

exports[`triggerGenerator module if the generated memory map matches with the snapshot 1`] = `
Map {
  "counter" => Object {
    "byteAddress": 0,
    "default": 0,
    "id": "counter",
    "lineNumber": 4,
    "relativeWordAddress": 0,
    "type": 5,
    "wordSize": 1,
  },
  "rate" => Object {
    "byteAddress": 4,
    "default": 0,
    "id": "rate",
    "lineNumber": 5,
    "relativeWordAddress": 1,
    "type": 6,
    "wordSize": 1,
  },
  "out" => Object {
    "byteAddress": 8,
    "default": 0,
    "id": "out",
    "lineNumber": 6,
    "relativeWordAddress": 2,
    "type": 4,
    "wordSize": 1,
  },
  "defaultValue" => Object {
    "byteAddress": 12,
    "default": 0,
    "id": "defaultValue",
    "lineNumber": 7,
    "relativeWordAddress": 3,
    "type": 5,
    "wordSize": 1,
  },
  "reset" => Object {
    "byteAddress": 16,
    "default": 12,
    "id": "reset",
    "lineNumber": 8,
    "relativeWordAddress": 4,
    "type": 3,
    "wordSize": 1,
  },
}
`;

exports[`triggerGenerator module if the wat code matches with the snapshot 1`] = `
"(module
  (type (;0;) (func))
  (import \\"js\\" \\"memory\\" (memory (;0;) 1))
  (func (;0;) (type 0)
    (local i32 i32 i32)
    i32.const 8
    i32.load
    local.set 1
    i32.const 0
    i32.load
    local.set 0
    i32.const 4
    i32.load
    local.set 2
    i32.const 16
    i32.load
    i32.load
    i32.const 0
    i32.gt_s
    if  ;; label = @1
      i32.const 0
      local.set 0
      i32.const 0
      local.set 1
    end
    local.get 1
    i32.const 32767
    i32.ge_s
    if  ;; label = @1
      i32.const 0
      local.set 1
    end
    local.get 0
    local.get 2
    i32.ge_s
    if (result i32)  ;; label = @1
      i32.const 32767
      local.set 1
      i32.const 0
    else
      local.get 0
      i32.const 1
      i32.add
    end
    local.set 0
    i32.const 8
    local.get 1
    i32.store
    i32.const 0
    local.get 0
    i32.store)
  (export \\"test\\" (func 0)))
"
`;
